test_conformance/run_conformance.py

```diff
@@ -206,75 +210,90 @@ def run_test_checking_output(current_directory, test_dir, log_file):
-        sys.stdout.flush()
-
-        # Run the test
-        result = 0
-        start_time = time.time()
-        try:
-            process_pid = 0
-            result = run_test_checking_output(current_directory, test_dir, log_file)
-        except KeyboardInterrupt:
-            # Catch an interrupt from the user
-            write_screen_log("\nFAILED: Execution interrupted.  Killing test process, but not aborting full test run.")
-            os.kill(process_pid, 9)
-            if sys.version_info[0] < 3:
-                answer = raw_input("Abort all tests? (y/n)")
-            else:
-                answer = input("Abort all tests? (y/n)")
-            if answer.find("y") != -1:
-                write_screen_log("\nUser chose to abort all tests.")
-                log_file.close()
-                sys.exit(-1)
-            else:
-                write_screen_log("\nUser chose to continue with other tests. Reporting this test as failed.")
-                result = 1
-        run_time = (time.time() - start_time)
-
-        # Move print the finish status
-        if result == 0:
-            print("(" + get_time() + ")     PASSED " + test_name.ljust(40) + ": (" + str(int(run_time)).rjust(3) + "s, test " + str(test_number).rjust(3) + os.sep + str(len(tests)) + ")", end='')
-        else:
-            print("(" + get_time() + ")     FAILED " + test_name.ljust(40) + ": (" + str(int(run_time)).rjust(3) + "s, test " + str(test_number).rjust(3) + os.sep + str(len(tests)) + ")", end='')
-
-        test_number = test_number + 1
-        log_file.write("     ----------------------------------------------------------------------------------------\n")
-        log_file.flush()
-
-        print("")
-        if result != 0:
-            log_file.write("  *******************************************************************************************\n")
-            log_file.write("  *  (" + get_time() + ")     Test " + test_name + " ==> FAILED: " + str(result) + "\n")
-            log_file.write("  *******************************************************************************************\n")
-            failures = failures + 1
-        else:
-            log_file.write("     (" + get_time() + ")     Test " + test_name + " passed in " + str(run_time) + "s\n")
-
-        log_file.write("     ----------------------------------------------------------------------------------------\n")
-        log_file.write("\n")
+    run_in_parallel = False
+    # Open the log file
+    for arg in sys.argv:
+        match = re.search("RUN_PARALLEL", arg)
+        if match:
+            run_in_parallel = True
+    if run_in_parallel:
+        pool = multiprocessing.Pool(4)
+        failures, previous_test, test_number = pool.starmap(run_specific_test, zip(tests, repeat([failures, previous_test, test_number])))
+    else:
+        for test_itr in tests:
+            failures, previous_test, test_number = run_specific_test(test_itr, [failures, previous_test, test_number])
```

Comment:
From the code, to run in parallel, the word "RUN_PARALLEL" is expected in the arg list. But this causes errors right at the beginning because it is not ignored like other args in lines 337-340 (in changed file).
So the following should be added right after:

if re.search("RUN_PARALLEL", arg):
        continue 

Second,  after the above is fixed, it still can't run (at least in windows). Following errors are seen: "An attempt has been made to start a new process before the current process has finished its bootstrapping phase."
This can be fixed by wrapping the calling of the main execution function (run_tests) and maybe surrounding code in a "main" function and then calling it as below:

if __name__=="__main__":
    main() 

After this, the tests run to completion (for a small subset of tests (6)) but "ValueError: too many values to unpack" is seen at the end.

The test specific stdout is also all mixed up due to tests running in parallel.

The help message should also be updated to mention the new "RUN_PARALLEL" argument

Looks like change needs some re-work before it can be reviewed again.

Reply:
Can we remove the focused review tag on this one if we are planning a different fix?